### DISCUSSION TITLE
Map tab to react to series/units being changed

### DISCUSSION THREAD

**Discussion by LucyGwilliamAdmin** [open-sdg/open-sdg#1456] [Ideas]
**Is your feature request related to a problem? Please describe.**
If an indicator has a map and more than one series, the map tab doesn't react to the series being changed. However, the title changes.

This can be misleading as users could think that the map is showing data for the series but it's not (unless it's the first series)

**Describe the solution that you need**
When the series is changed, the map should change to reflect the changes in data

**Why do you need this feature?**
So that users aren't misled in thinking that the map is showing something it isn't

**Describe alternatives you've considered**
Depending on how long this may take and difficulty, maybe something quicker/easier could be implemented in short term e.g. that makes it clear that the map is showing the first series.

**Comment by LucyGwilliamAdmin** [comment]
@EmmaWoodONS I've just added this issue based on what you were saying about maps and series but feel free to make some changes to the description if I've missed anything

**Comment by EmmaWoodONS** [comment]
Thanks Lucy - I think that covers it. I just have one slight addition:

Even if the start_value is set to a series that is not the default, the title of the map changes to that for the start_value, but the map itself stays on the default first series.

**Comment by WohnungNDS** [comment]
@LucyGwilliamAdmin any progress in here yet?
in my case the map title doesnot change at all.... 
I have looked through the issues seeking for an option to change the map titel...

**Comment by brockfanning** [comment]
Adding a bit of history here: In the past we have avoided any connection between the sidebar disaggregation controls and the map, because it is impossible to display more than one disaggregation from the same group on the map. For example, it would be impossible for a map to show both "Male" and "Female" on one map. There is a long-standing pull-request (#663) where we attempted to have a drop-down on the map to allow switching between all the possible disaggregation combinations. But it was still fundamentally disconnected from the sidebar disaggregation controls.

That said, the idea of connecting the map to the "Series" toggle specifically, may be possible, because of the fact that the Series toggle is a radio button, which can only have one selected value at a time. The "Unit" toggle could also affect it, for that matter, since it is also a radio button. So, I think we could consider connecting the "Series" and "Unit" toggles to the map.

But we would not be able to connect the disaggregation drop-downs, because of the multiple-disaggregation issue mentioned above. So we would need to think about user experience. Right now, we completely hide the left sidebar when the map is active. If it were connected to the Series/Unit, then presumably the sidebar would need to stay open. But then, the disaggregation drop-downs would be visible, which may make the users think that they are affecting the map too.

**Comment by brockfanning** [comment]
More updates on this:

We have tried a proof-of-concept where the left sidebar stays visible, and the toggles for unit and series affect the map, but the other disaggregation controls disappear. A problem with this approach is that the behavior of the left sidebar changes, but the left sidebar is not inside the "tab pane". In other words, the left sidebar works differently depending on whether the "Map" tab is active or not. This breaks a best-practice of tabs: everything that is specific to the active tab should be inside the tab pane. We have considered moving the sidebar inside the tab panes, but noted that it would be problematic to do this, for example, because it would mean that selections made while on the "Chart" tab would not carry over to the "Table" tab -- and switch between Chart and Table to see the same data is an important feature of Open SDG.

Another approach we could consider would be to duplicate the series/unit toggles on the map itself (like in the upper-left corner).

**Comment by christoph-LSN** [comment]
@brockfanning @LucyGwilliamAdmin  is there any progress in this issue?

We want to get online with the migration report in the end of july. So I have to decide what to do with my data. I definitly have to show some maps. So I am thinking about to delete all GeoCodes despite the current year and one of the series....

e.g here [https://christoph-lsn.github.io/MT_Site/3-1-2/](https://christoph-lsn.github.io/MT_Site/3-1-2/)

I will only have Geocodes for the Year 2021 and in "Kinderbetreuung" "in der Familie wird vorrangig nicht deutsch gesprochen"

Or will it be possible to display more data on the map?

[siterepo](https://github.com/christoph-LSN/MT_Site)
[datarepo](https://github.com/christoph-LSN/IM_Data)

**Reply by brockfanning** [reply]
@christoph-LSN It's a very rough prototype would surely be different by the time it is released, but I would welcome any feedback on this branch: brockfanning/open-sdg@map-disaggregation-in-leaflet

This should add disaggregation controls inside the map itself, which is an approach that seems to avoid the problems that we've been running into. Note that this branch includes all the latest code from 2.0.0-dev.

**Comment by christoph-LSN** [comment]
@brockfanning 
@LucyGwilliamAdmin 

I have tried it as you can see [here](https://christoph-lsn.github.io/MT_Site/3-1-3/)

I changed also the requrements.txt like this:

`git+https://github.com/open-sdg/sdg-build@2.0.0-dev`

and the gemfile
`gem 'jekyll-open-sdg-plugins', git: 'https://github.com/open-sdg/jekyll-open-sdg-plugins.git', branch: '2.0.0-dev'`

set 
`bootstrap:5:true `
and 
`chart_3.js: true`

The graph does not work anymore and the table is not shown anymore....

The map looks like before there are no options on the map

did I something wrong?

siterepo: [https://github.com/christoph-LSN/MT_Site](https://github.com/christoph-LSN/MT_Site)
datarepo; [https://github.com/christoph-LSN/IM_Data](https://github.com/christoph-LSN/IM_Data)

**Reply by brockfanning** [reply]
@christoph-LSN It's difficult to say what might be going wrong, because you have a large amount of overridden files in your site repository. It looks like maybe you're overriding _all_ the open-sdg files, or at least most of them. That will definitely interfere with the changes in my branch. To help me troubleshoot it would be helpful if you could provide a version of your site repo with only the minimum of overrides. That way I can more easily pinpoint any bugs in my code.

**Reply by christoph-LSN** [reply]
@brockfanning I have not so many overriden files... but the repo is quite old (two years or so)... Anyway today I try to use the current site-starteter repo for testing.....

**Comment by christoph-LSN** [comment]
@brockfanning Today I used the data-starter and site starter repo to create two new repos for testing.

[site repo](https://github.com/christoph-LSN/MT_Test)
[data repo](https://github.com/christoph-LSN/MT_Data_Test)

I only added the bootstrap5 files, geojson files and some data from my old data repo.

Using bootstrap5 and chart3.js makes the table, chart and map don't work anymore.

Without bootstrap5 and chart3.js  everything works but there are no dissaggregation controls on the map

Exampel indicator: https://christoph-lsn.github.io/MT_Test/1-1-1/

![image](https://user-images.githubusercontent.com/94443931/171232615-ef039483-9405-4fa8-93e1-089849df7b5e.png)

Changing the gemfile to 2.0.0-dev shows this error:

`  * jekyll-open-sdg-plugins-1.7.0.pre.beta1
  * jekyll-open-sdg-plugins-1.7.0
  * jekyll-open-sdg-plugins-1.8.0.pre.beta1
  * jekyll-open-sdg-plugins-1.8.0
Error: Process completed with exit code 7.`

**Reply by brockfanning** [reply]
@christoph-LSN I should clarify what I mean by an "override" - any file that is in the _includes folder or the _layouts folder -- if it is named the same as a file in the Open SDG _includes and _layouts folders -- is an "override".

In order to get a clean install you should preferably have no overrides. I can get your MT_Test site to work by making these changes:
1. Delete everything in _includes
2. Delete everything in _layouts
3. Change the Gemfile to the 2.0.0-dev branch of jekyll-open-sdg-plugins, like this: `gem 'jekyll-open-sdg-plugins', git: 'https://github.com/open-sdg/jekyll-open-sdg-plugins.git', branch: '2.0.0-dev'`

When I do this locally I'm able to see the map disaggregation on 1.1.1. I'm already seeing a problem, that it does not translate the disaggregations. But otherwise it appears functional.

**Comment by christoph-LSN** [comment]
@brockfanning @LucyGwilliamAdmin @MoBosse  

Thank you! It works. e.g. here: https://christoph-lsn.github.io/MT_Site/1-2-3/
I have changed it in my [MT_Site](https://github.com/christoph-LSN/MT_Site) repo 

one thing I have seen is, that the disaggregations are not sortet like in the chart oder table view

![image](https://user-images.githubusercontent.com/94443931/171373078-e5ad3404-c613-47bb-865d-9eff18a15fe4.png)

Another thing I observed is that the legend entries are changing too. That's great. But something is wrog with the text and value spaces in the legend. But may be this is a more gerneal thing and has less to do with the disaggregations

![image](https://user-images.githubusercontent.com/94443931/171373680-ffcbc7d0-8250-497f-9d86-76119e67674e.png)
In General this is what I need!

This "overrides" are left from my test with bootstrap5 and chart3.js. I haven't seen this bootstrap5 files in your open-sdg repo, so I thought I have to copy them from the other open-sdg/open-sdg repo

Thank you Brock!

**Reply by brockfanning** [reply]
Thanks for the feedback @christoph-LSN. I'll look into the sorting issue, and try to make sure things are getting translated.

For your awareness - when upgrading from 1.8.0 to 2.0.0, the "bootstrap5" subfolders will no longer be necessary. This is because in 2.0.0 Bootstrap 5 is automatically used (there is no option) so there is no reason to have separate include files and layouts. This is briefly mentioned [here](https://github.com/open-sdg/open-sdg/blob/2.0.0-dev/docs/upgrades/upgrading-2-0-0.md#bootstrap_5---always-on) but we will be posting a much more detail blog post about it before releasing 2.0.0.

**Reply by MoBosse** [reply]
This looks like a good approach to me. Definitely something we would use as well.
At first glance, I wonder how you could select the headline time series here. For example, in @WohnungNDS 's example, "Geschlecht" = "Insgesamt" (=total) is selectable, but what if the "Geschlecht" column in the data file was empty?

Also, there seems to be a bug somewhere: in IE the data is not loaded correctly.

**Reply by brockfanning** [reply]
@MoBosse I believe right now, if there is an empty disaggregation value, it appears as a blank option. This is not ideal and I think we would need to come up with a better label, like "Total", "All", "Aggregated", etc.

Regarding IE, an important point: in version 2.0.0 support for IE will be dropped. This is because 2.0.0 will use Bootstrap 5, and Bootstrap 5 does not support IE. I can take a look at the javascript in this PR and see if there is anything used that would be a problem for IE. But in general we won't be guaranteeing IE support in 2.0.0 and higher.

**Comment by christoph-LSN** [comment]
@brockfanning @MoBosse 

Another point: If there is no disaggregation in the data the change breakkdown buttom should be hidden

https://christoph-lsn.github.io/MT_Site/1-2-1/

![image](https://user-images.githubusercontent.com/94443931/171577429-b705c0df-32c0-42fa-b90d-26d0667574c2.png)
